rust_src/src/lread.rs

```diff
@@ -155,6 +159,41 @@ pub unsafe fn defvar_per_buffer_offset(
     }
 }
 
+/// Read a byte from stdio. If it has lookahead, use the stored value.
+/// If read over network is interrupted, keep trying until read succeeds.
+#[no_mangle]
+pub unsafe extern "C" fn readbyte_from_stdio() -> i32 {
+    // If infile has lookahead, use stored value
+    if (*infile).lookahead > 0 {
+        let idx = (*infile.sub(1)).lookahead as usize;
+        return (*infile).buf[idx].into();
+    }
+
+    let instream = (*infile).stream;
+
+    block_input();
+
+    // Interrupted read have been observed while reading over the network.
+    let mut c;
```

Comment:
An alternative implementation that might be more idiomatic Rust is:

    loop {
        c = getc_unlocked(instream);
        if c == libc::EOF && errno().0 == libc::EINTR && ferror_unlocked(instream) > 0 {
            unblock_input();
            maybe_quit();
            block_input();
            clearerr_unlocked(instream);
        } else {
            break;
        }
    }

If we factor out the check....

    fn is_interrupted_read(stream: TheStreamType, value: i32) -> bool {
        c == libc::EOF && errno().0 == libc::EINTR && ferror_unlocked(stream) > 0
    }

    loop {
        c = getc_unlocked(instream);
        if not is_interrupted_read(instream, c) {
            break;
        }
        unblock_input();
        maybe_quit();
        block_input();
        clearerr_unlocked(instream);
    }

hmmm

    fn interruptable_read(stream: TheStreamType) -> i32 {
        loop {
            let c = getc_unlocked(stream);
            if c == libc::EOF && errno().0 == libc::EINTR && ferror_unlocked(instream) > 0 {
                unblock_input();
               maybe_quit();
               block_input();
               clearerr_unlocked(instream);
           } else {
                return c;
           }
        }
    }

    let c = interruptable_read(instream);
    
    unblock_input();

    if c != libc::EOF {
        c
    } else {
        -1
    }

Reply:
The loop is definitely a more readable approach, but I'm inclined to leave the code in the function body as it's not used anywhere else, and I consider the slight readability trade-off worth having the code in one place.

Reply:
Fair enough.